### PR TITLE
Allow passing through prototypes to skip merge deep for

### PR DESCRIPTION
**Pull request by francescoferraioli**
Yet to create a PR in TypeORM for this one.

Question: Constructor or Prototype?

A part of me is not super happy with this code and I'd love to clean it up somehow. It's just a bit hard because the only configuration point is at the connection level and the problematic code is in a static method so we have to just pass in the parameter for the state. I also had to make it the first parameter because of the use of the rest operator on the last parameter (...sources: any[]).

I was thinking that maybe a nicer and more flexible approach could be to instead of passing in a list of constructors/prototypes, we could pass in a list of a predicate and a mergeDeep. Interested in your thoughts.

It also doesn't have tests which I'd need to add for the TypeORM PR. Do we want it for our own repo too? Maybe we might as well?

### CODE REVIEW

File: src/metadata/EntityMetadata.ts [hunk 1]
```diff
@@ -590,7 +590,7 @@ export class EntityMetadata {
         if (!entity)
             return undefined;
 
-        return EntityMetadata.getValueMap(entity, this.primaryColumns, { skipNulls: true });
+        return this.getValueMap(entity, this.primaryColumns, { skipNulls: true });
```
Comment (Benno007) [inline src/metadata/EntityMetadata.ts:593]: Too much to leave this static and pass the props in as options, or another param here? I'm not sure if it matters though, pleerock might prefer the static method

File: src/persistence/subject-builder/ManyToManySubjectBuilder.ts [hunk 2]
```diff
@@ -232,10 +232,10 @@ export class ManyToManySubjectBuilder {
 
         const identifier: ObjectLiteral = {};
         relation.junctionEntityMetadata!.ownerColumns.forEach(column => {
-            OrmUtils.mergeDeep(identifier, column.createValueMap(column.referencedColumn!.getEntityValue(ownerEntityMap)));
+            OrmUtils.mergeDeep(subject.metadata.connection.options.skipMergeDeepPrototypes, identifier, column.createValueMap(column.referencedColumn!.getEntityValue(ownerEntityMap)));
         });
         relation.junctionEntityMetadata!.inverseColumns.forEach(column => {
-            OrmUtils.mergeDeep(identifier, column.createValueMap(column.referencedColumn!.getEntityValue(inverseEntityMap)));
+            OrmUtils.mergeDeep(subject.metadata.connection.options.skipMergeDeepPrototypes, column.createValueMap(column.referencedColumn!.getEntityValue(inverseEntityMap)));
```
Comment (Benno007) [inline src/persistence/subject-builder/ManyToManySubjectBuilder.ts:238]: missing identifier?

File: src/util/OrmUtils.ts [hunk 3]
```diff
@@ -82,17 +83,18 @@ export class OrmUtils {
                 && !(value instanceof Set)
                 && !(value instanceof Date)
                 && !(value instanceof Buffer)
-                && !(value instanceof RegExp)) {
+                && !(value instanceof RegExp)
+                && skipMergeDeepPrototypes.every(X => !(value instanceof X))) {
```
Comment (Benno007) [inline src/util/OrmUtils.ts:87]: So given this is a constructor, would we just be using it on like whatever it was in ZonedDateTime that's circular? ZoneOffset type or something?

Wondering on things like:
const a = { name: 'hello' }
a.prop = a
Would people be able to block on `a.prop`? Probably not with instanceof I guess. I'm not sure if that's a problem you can solve by using prototypes over constructors? I read that mergedeep issue, you'd have to maintain a map of references or something? I think this is ok for the scope of this task.

I think ideally, some new work for the tech debt backlog could be using native dates, so we don't need a custom fork for another 3 years :p

**Comment by Benno007** [comment]
I think some tests would be a good idea if you can manage it

**Comment by danewilson** [comment]
If you're not convinced this change would make it into the main repo, and we're planning on moving away from the problematic type anyway, should we just hardcode the problematic type(s) in our fork and then remove it once we're no longer using that type(s)?

**Comment by francescoferraioli** [comment]
> If you're not convinced this change would make it into the main repo, and we're planning on moving away from the problematic type anyway, should we just hardcode the problematic type(s) in our fork and then remove it once we're no longer using that type(s)?

@danewilson - the problematic type is js-joda's ZoneOffset. We can't hardcode it as it's not a dependency of TypeORM.

**Comment by danewilson** [comment]
> > If you're not convinced this change would make it into the main repo, and we're planning on moving away from the problematic type anyway, should we just hardcode the problematic type(s) in our fork and then remove it once we're no longer using that type(s)?
> 
> @danewilson - the problematic type is js-joda's ZoneOffset. We can't hardcode it as it's not a dependency of TypeORM.

Understood, but it could be a (peer/dev)dependency of our fork? You could also (super hacky) `x.constructor.name === 'ZoneOffset'. Might make our fork easier to maintain if we don't think this change would ever be merged to mainline.

**Comment by francescoferraioli** [comment]
> Understood, but it could be a (peer/dev)dependency of our fork? You could also (super hacky) `x.constructor.name === 'ZoneOffset'. Might make our fork easier to maintain if we don't think this change would ever be merged to mainline.

That is so true! Very interesting. It certainly would make our fork easier to maintain but completely leave us doomed to this fork until we move away from js-joda.

I like the idea of putting a PR up into the mainline as it will help out other people as well, and maybe I still can, just not in our fork.

It's an interesting option and leaves us with an interesting choice. I am actually leaning towards doing the hacky solution for our fork and maybe putting a PR into the mainline for the better solution and see what comes from it.

**Comment by francescoferraioli** [comment]
Closing in favour of https://github.com/FTRLabs/typeorm/pull/14

**Pull request closed without merging**